File: MetaGeta/MetaGeta.DataStore/GlobalSettings/GlobalSetting.vb
Public Class GlobalSetting
    Implements INotifyPropertyChanged, IDataErrorInfo
    Private ReadOnly m_Name As String
    Private ReadOnly m_DefaultValue As String
    Private ReadOnly m_Type As GlobalSettingType

    Private m_Value As String

    Public Sub New(ByVal name As String, ByVal value As String, ByVal defaultValue As String, ByVal type As GlobalSettingType)
        m_Name = name

        m_DefaultValue = defaultValue
        m_Type = type

        If value = m_DefaultValue Then
            m_Value = Nothing
        Else
            m_Value = value
        End If
    End Sub

    Public ReadOnly Property Name() As String
        Get
            Return m_Name
        End Get
    End Property
    Public ReadOnly Property DefaultValue() As String
        Get
            Return m_DefaultValue
        End Get
    End Property
    Public ReadOnly Property Type() As GlobalSettingType
        Get
            Return m_Type
        End Get
    End Property
    Public Property Value() As String
        Get
            If m_Value Is Nothing Then
                Return m_DefaultValue
            Else
                Return m_Value
            End If
        End Get
        Set(ByVal value As String)
            If value <> m_Value Then
                If value = m_DefaultValue Then
                    m_Value = Nothing
                Else
                    m_Value = value
                End If
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs("Value"))
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs("IsDefault"))
            End If
        End Set
    End Property
    Public ReadOnly Property ValueOrNull() As String
        Get
            Return m_Value
        End Get
    End Property
    Public ReadOnly Property IsDefault() As Boolean
        Get
            Return m_Value = Nothing
        End Get
    End Property


    Public Event PropertyChanged(ByVal sender As Object, ByVal e As System.ComponentModel.PropertyChangedEventArgs) Implements System.ComponentModel.INotifyPropertyChanged.PropertyChanged

    Public ReadOnly Property [Error]() As String Implements IDataErrorInfo.Error
        Get
            Return Nothing
        End Get
    End Property

    Default Public ReadOnly Property Item(ByVal propertyName As String) As String Implements System.ComponentModel.IDataErrorInfo.Item
        Get
            If propertyName <> "Value" Then Return Nothing

            Select Case m_Type
                Case GlobalSettingType.Number
                    Dim v As Double
                    If Double.TryParse(Value, v) Then
                        Return Nothing
                    Else
                        Return String.Format("""{0}"" is not a number.", Value)
                    End If
                Case GlobalSettingType.Directory
                    Return If(Directory.Exists(Environment.ExpandEnvironmentVariables(Value)), Nothing, String.Format("Can't find directory ""{0}"".", Value))
                Case GlobalSettingType.File
                    Return If(File.Exists(Environment.ExpandEnvironmentVariables(Value)), Nothing, String.Format("Can't find file ""{0}"".", Value))
                Case GlobalSettingType.LongText
                    Return Nothing
                Case GlobalSettingType.ShortText
                    Return Nothing
                Case Else
                    Throw New Exception()
            End Select
        End Get
    End Property

    Public ReadOnly Property IsValid() As Boolean
        Get
            Return Item("Value") Is Nothing
        End Get
    End Property
End Class

File: MetaGeta/MetaGeta.DataStore/GlobalSettings/GlobalSetting.cs
// Copyright 2009 Will Thomas
// 
// This file is part of MetaGeta.
// 
// MetaGeta is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// MetaGeta is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MetaGeta. If not, see <http://www.gnu.org/licenses/>.

#region

using System;
using System.ComponentModel;
using System.IO;

#endregion

namespace MetaGeta.DataStore {
    public class GlobalSetting : INotifyPropertyChanged, IDataErrorInfo {
        private readonly string m_DefaultValue;
        private readonly string m_Name;

        private readonly GlobalSettingType m_Type;

        private string m_Value;

        public GlobalSetting(string name, string value, string defaultValue, GlobalSettingType type) {
            m_Name = name;

            m_DefaultValue = defaultValue;
            m_Type = type;

            if (value == m_DefaultValue)
                m_Value = null;
            else
                m_Value = value;
        }

        public string Name {
            get { return m_Name; }
        }

        public string DefaultValue {
            get { return m_DefaultValue; }
        }

        public GlobalSettingType Type {
            get { return m_Type; }
        }

        public string Value {
            get {
                if (m_Value == null)
                    return m_DefaultValue;
                else
                    return m_Value;
            }
            set {
                if (value != m_Value) {
                    if (value == m_DefaultValue)
                        m_Value = null;
                    else
                        m_Value = value;
                    if (PropertyChanged != null)
                        PropertyChanged(this, new PropertyChangedEventArgs("Value"));
                    if (PropertyChanged != null)
                        PropertyChanged(this, new PropertyChangedEventArgs("IsDefault"));
                }
            }
        }

        public string ValueOrNull {
            get { return m_Value; }
        }

        public bool IsDefault {
            get { return m_Value == null; }
        }

        public bool IsValid {
            get { return this["Value"] == null; }
        }

        #region IDataErrorInfo Members

        public string Error {
            get { return null; }
        }

        public string this[string propertyName] {
            get {
                if (propertyName != "Value")
                    return null;

                switch (m_Type) {
                    case GlobalSettingType.Number:
                        double v = 0;
                        if (double.TryParse(Value, out v))
                            return null;
                        else
                            return string.Format("\"{0}\" is not a number.", Value);
                        break;
                    case GlobalSettingType.Directory:
                        return Directory.Exists(Environment.ExpandEnvironmentVariables(Value)) ? null : string.Format("Can't find directory \"{0}\".", Value);
                    case GlobalSettingType.File:
                        return File.Exists(Environment.ExpandEnvironmentVariables(Value)) ? null : string.Format("Can't find file \"{0}\".", Value);
                    case GlobalSettingType.LongText:
                        return null;
                    case GlobalSettingType.ShortText:
                        return null;
                    default:
                        throw new Exception();
                }
            }
        }

        #endregion

        #region INotifyPropertyChanged Members

        public event PropertyChangedEventHandler PropertyChanged;

        #endregion
    }
}

namespace MetaGeta.DataStore {
    public enum GlobalSettingType {
        ShortText,
        LongText,
        Number,
        Directory,
        File
    }
}